### PR TITLE
Split AOT Engine Runtime

### PR DESCRIPTION
**Pull request by GaryQian**
## Description

Adds a Dart_DeferredLibraryHandler to plumb `loadLibrary()` calls from dart through the dart isolate, platform configuration, runtime controller into the shell entrypoint of Engine::RequestDartDeferredLibrary.

Also connects the shell's Engine::LoadDartDeferredLibrary to the dart isolate.

## Related Issues

https://github.com/flutter/flutter/issues/62229
https://github.com/flutter/flutter/issues/57617

### CODE REVIEW

File: runtime/dart_isolate.cc
Comment (chinmaygarde) [inline runtime/dart_isolate.cc:343]: Here and elsewhere, `tonic::LogIfError` unless there is a reason this can only be an API error.

File: runtime/dart_isolate.cc
Comment (chinmaygarde) [inline runtime/dart_isolate.cc:323]: Unchecked error handle.

File: shell/common/engine.cc
Comment (chinmaygarde) [inline shell/common/engine.cc:512]: Can this return an error (or boolean to indicate one) if the request failed. I am not comfortable swallowing a potential error here.

File: runtime/runtime_controller.h
Comment (chinmaygarde) [inline runtime/runtime_controller.h:485]: Please make a note here about the lifecycle of the shared library handle. Is it tied to the isolate or the VM for example. Also, make a note of who owns it in the engine right now.

File: runtime/runtime_controller.h
Comment (chinmaygarde) [inline runtime/runtime_controller.h:504]: I think, at this layer, we can use `fml::Mapping`s instead. We'll still have to use raw pointers to talk to Dart. Using mappings (specifically `SymbolMapping`s) will make the code self documenting and readers will be better able to follow ownership.

File: runtime/runtime_controller.h
Comment (chinmaygarde) [inline runtime/runtime_controller.h:517]: The docs say you return something by you actually return void. I think you missed propagating the error.

**Comment by chinmaygarde** [comment]
Is this still WIP? Its causing it to skip review queues.

**Comment by GaryQian** [comment]
I had it marked WIP when I last pushed, it should go through precheck whenever I push again.

File: shell/common/engine.cc
Comment (GaryQian) [reply shell/common/engine.cc:512]: The idea is that errors/failures should invoke `LoadDartDeferredLibraryFailure`. Capturing/returning any errors here may inadvertently indicate to the dev that the error handling is complete. This method is meant to inform the embedder/engine that the dart lib is requested, but makes no promises that it will or even should succeed.

I can make sure I document this.

Comment (GaryQian) [reply shell/common/engine.cc:512]: The full LoadDartDeferredLibraryFailure implementation is coming soon in a future PR.

Comment (GaryQian) [reply shell/common/engine.cc:512]: Also, this request is not blocking, as the majority of the work happens async in the downloading/extracting/opening/io of the lib. Thus, even a return value of true would not indicate success since most of the work has yet to be done.

File: shell/common/engine.h
Comment (xster) [inline shell/common/engine.h:801]: If you're being explicit about taking a `SymbolMapping`, should the signature be specified to take `SymbolMapping` too?

**Review by xster — APPROVED** [review]
Please wait for Chinmay's LGTM too. But LGTM

File: shell/common/engine.h
Comment (chinmaygarde) [reply shell/common/engine.h:801]: `SymbolMapping` is used when we own the the native library handle too. In some cases (usually where the platform or embedder owns this handle), we can use an `UnownedMapping`. In any case, both are subclasses of `fml::Mapping`.  I don't think this layer should care how the mapping is resolved. Just that there is a mapping whose lifecycle has been sorted out (and readers can follow).

File: runtime/dart_isolate.cc
Comment (chinmaygarde) [inline runtime/dart_isolate.cc:340]: A subtle error has been introduced here when you went from the C++ API to the Dart C API. The symbol mapping owns the library handle you created later in this patch. However, both go out of scope here. The library is now un-reffed and may close the library handle. This can cause issues depending on the dynamic linker implementation. From the docs of `dlcose`:

```
"The use of dlclose() reflects a statement of intent on the part of the process,
but does not create any requirement upon the implementation, such as removal
of the code or symbols referenced by handle. Once an object has been closed
using dlclose() an application should assume that its symbols are no longer
available to dlsym()."
```

Depending on the implementation, subject to timing issues within the implementation, and, other references to this library by the engine, the OS is also free to unmap the memory you just gave the Dart code via the C API which would then lead to segfaults.

We do this in other places by owning a reference to a `flutter::DartSnapshot` which is nothing but two mappings. I suggest you do the same. Maybe toss it inside a `set` in `DartIsolate`.

I am glad we are running into these design issues now. Using the C++ wrappers forces you to think about lifecycle issues that raw pointers don't. This would be a very hard to debug segfault otherwise.

**Review by chinmaygarde — CHANGES_REQUESTED** [review]
Please don't land this till the symbol lifecycle issue is fixed. Also, this has no tests but I am not sure if this is WIP. When you do, please add one to `runtime_unittests`.

File: runtime/dart_isolate.cc
Comment (GaryQian) [reply runtime/dart_isolate.cc:340]: Ahh, I see what you meant by lifecycle now. Thanks for catching this and the explanation.

Comment (chinmaygarde) [reply runtime/dart_isolate.cc:340]: Right. When you vend a raw pointer, you need to know upfront till when the user expects to use the pointer. That way, you can keep what is being pointed to alive for at least that long.

From my last comment on lifecycle, it is still not clear to me if the VM can use this post isolate termination. Presumably not, but just good to check.

**Comment by GaryQian** [comment]
Some test behavior is waiting on https://dart-review.googlesource.com/c/sdk/+/174563 to land to better handle failure to load cases.

File: testing/elf_loader.h
Comment (chinmaygarde) [inline testing/elf_loader.h:18]: Not sure how this name was generated. I see the fixture generator specifying the app_elf_snapshot base name but unclear about the `so-2.part` bit. Can you add a comment here for readers on how the tooling generates this name? Or, just be explicit about the name if the capability to do so is already present.

File: testing/fixture_test.h
Comment (chinmaygarde) [inline testing/fixture_test.h:11]: Nit: Unsure why this include was necessary. ELFAOTSymbols is not in fml/mapping.

File: runtime/runtime_controller.h
Comment (chinmaygarde) [inline runtime/runtime_controller.h:533]: Nit: The doc generation works better when the docs are added to the virtual methods in the base class instead of the overriden ones in the subclass. So, I'd move these doc comments to `platform_configuration.h`. Keep the `| PlatformConfigurationClient|` comment though.

File: runtime/runtime_controller.h
Comment (chinmaygarde) [inline runtime/runtime_controller.h:510]: I thought you were moving these over to use `flutter::DartSnapshot`. Either way is fine though as the DartSnapshot is nothing but a pointer to two mappings. I see you are wrapping this up in the implementation. you could also just make the caller wrap it for you. Just a thought.

File: runtime/runtime_controller.cc
Comment (chinmaygarde) [inline runtime/runtime_controller.cc:7]: I believe this include is spurious.

File: runtime/dart_isolate.cc
Comment (chinmaygarde) [inline runtime/dart_isolate.cc:1046]: Shouldn't we be returning an error handle here instead of `Dart_Null`?

**Review by chinmaygarde — APPROVED** [review]
Except the null return in case of error in `DartIsolate::OnDartLoadLibrary`, everything else is nits and minor observations. Otherwise LGTM. Nicely done.

File: testing/elf_loader.h
Comment (GaryQian) [reply testing/elf_loader.h:18]: Yeah, sure. This naming is just what gen_snapshot defaults to. The `2` indicates the id for the loading unit is 2 (base module is 1).

**Comment by aam** [comment]
I believe the test DartIsolateTest_ValidLoadingUnitSucceeds_Test introduced in this CL is flaky: https://github.com/flutter/flutter/issues/71661

**Pull request merged**
